sklearn/feature_selection/univariate_selection.py

```diff
@@ -226,6 +226,185 @@ def chi2(X, y):
+    selection in text categorization. Proceedings of ICML'97, pp. 412-420.
+    http://citeseer.ist.psu.edu/viewdoc/summary?doi=10.1.1.32.9956
+
+    .. [3] F. Sebastiani. 2002. Machine Learning in Automatic Text
+    Categorization. ACM Computing Surveys (CSUR).
+    http://nmis.isti.cnr.it/sebastiani/Publications/ACMCS02.pdf
+
+    """
+
+    X = check_array(X, accept_sparse='csr')
+    if np.any((X.data if issparse(X) else X) < 0):
+        raise ValueError("Input X must be non-negative.")
+
+    if not isinstance(X, csr_matrix) and not isinstance(X, coo_matrix):
+        raise ValueError("Input X must be a sparse matrix.")
+
+    Y = LabelBinarizer().fit_transform(y)
+    if Y.shape[1] == 1:
+        Y = np.append(1 - Y, Y, axis=1)
+
+    # counts
+
+    fc_count = safe_sparse_dot(Y.T, X)          # n_classes * n_features
+    total = fc_count.sum(axis=0).reshape(1, -1).sum()
+    f_count = X.sum(axis=0).reshape(1, -1)
+    c_count = (X.sum(axis=1).reshape(1, -1) * Y).T
+
+    # probs
+
+    f_prob = f_count / f_count.sum()
+    c_prob = c_count / float(c_count.sum())
+    fc_prob = fc_count / total
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
+        scores = _ig(fc_count, c_count, f_count, fc_prob, c_prob, f_prob, total)
+
+    return np.asarray(scores).reshape(-1), []
+
+
+def igr(X, y):
+    """Compute an Information Gain Ratio score for each feature in the data.
+
+    The score can be used to weight features by informativeness or select the
+    most informative features for training and evaluating a classifier.
+
+    Information Gain measures the number of bits of information obtained about
+    the presence or absence of a class by knowing the presence or absence of the
+    feature. Information Gain Ratio aims to overcome one disadvantage of IG
+    which is the fact that IG grows not only with the increase of dependence
+    between f and c, but also with the increase of the entropy of f. That is why
+    features with low entropy receive smaller IG weights although they may be
+    strongly correlated with a class. IGR removes this factor by normalizing IG
+    by the entropy of the class.
+
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape = (n_samples, n_features_in)
+        Sample vectors.
+
+    y : array-like, shape = (n_samples,)
+        Target vector (class labels).
+
+    Returns
+    -------
+    scores : array, shape = (n_features,)
+
+    See also:
+    ---------
+    * ig : Information Gain
+    """
+
+    def get_t(f_prob):
+        t = np.log2(f_prob)
+        t[~np.isfinite(t)] = 0
+        return np.multiply(-f_prob, t)
+
+    X = check_array(X, accept_sparse='csr')
+    if np.any((X.data if issparse(X) else X) < 0):
+        raise ValueError("Input X must be non-negative.")
+
+    if not isinstance(X, csr_matrix) and not isinstance(X, coo_matrix):
+        raise ValueError("Input X must be a sparse matrix.")
```

Comment:
The implementation should be different for dense arrays - there are incompatible shape errors when one tries to pass a dense array.